### ISSUE TITLE
Exporting LTR Model in SavedModel Format

### ISSUE THREAD

**Issue by nsalerni** [tensorflow/ranking#53]
Could we please get an example of how to save the produced model in https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_libsvm.ipynb in SavedModel format (ref: https://www.tensorflow.org/guide/saved_model). I've tried a bunch of variations of exporting from the TensorFlow API docs and GitHub examples with no luck.

```
feature_columns = example_feature_columns()
feature_spec = tf.feature_column.make_parse_example_spec(feature_columns)
export_fn = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
ranker.export_savedmodel("savedmodel", export_fn)
```

With the example above I get ```KeyError: NumericColumn(key='1', shape=(1,), default_value=(0.0,), dtype=tf.float32, normalizer_fn=None)```

Some guidance on this problem would be greatly appreciated.

**Comment by eggie5** [comment]
I'm actually working on documenting the E2E process here: http://www.eggie5.com/145-tensorflow-model-exporing-and-serving, but this is the gist of for a ranker w/ 1 integer input `iid`:

```
def example_serving_input_fn():
    iid = tf.feature_column.numeric_column("iid", dtype=tf.int64, shape=[1])
    feature_spec = tf.feature_column.make_parse_example_spec([iid])
    return tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
```

```
path = ranker.export_saved_model(
        "output/exports/parsing", example_serving_input_fn())
```

You can then inspect the model by using the CLI:

```
!saved_model_cli show \
    --dir . \
    --tag_set serve \
    --signature_def predict 
```

```
The given SavedModel SignatureDef contains the following input(s):
  inputs['examples'] tensor_info:
      dtype: DT_STRING
      shape: (-1)
      name: input_example_tensor:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['output'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, -1)
      name: groupwise_dnn_v2/accumulate_scores/truediv:0
Method name is: tensorflow/serving/predict
```

You can rank some documents using the run command of the CLI. This example will rank a batch of 3 documents:

```
!saved_model_cli run \
    --dir . \
    --tag_set serve \
    --signature_def predict \
    --input_examples 'examples=[{"iid":[1]},{"iid":[2]},{"iid":[3]}]'
```
```
Result for output key output:
[[0.72908896]
 [0.73187715]
 [0.39144227]]
```

You can also deserialize/load the model from disk in Python:

```
import tensorflow as tf
import numpy as np
from tensorflow.python.saved_model import loader

def serialize_example(feature0, feature1):
    """
    Creates a tf.Example message ready to be written to a file.
    """
    def _int64_feature(value):
      """Returns an int64_list from a bool / enum / int / uint."""
      return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))

    # Create a dictionary mapping the feature name to the tf.Example-compatible
    # data type.

    feature = {
      'iid': _int64_feature(feature0),
      'uid': _int64_feature(feature1),
    }

    # Create a Features message using tf.train.Example.

    example_proto = tf.train.Example(features=tf.train.Features(feature=feature))
    return example_proto.SerializeToString()
  
n_observations=50
iids= np.array([0,1,2])#np.random.randint(0, 10000, n_observations)
uids= np.array([0,1,2])#np.random.randint(0, 2000, n_observations)
serialized_examples = []
for i in range(len(iids)):
    serialized_example = serialize_example(iids[i], uids[i])
    serialized_examples.append(serialized_example)
    
inputs_feed_dict = {'input_example_tensor:0': serialized_examples}

tags=["serve"]
signature_def_key = "predict"
saved_model_dir = "."

    
with tf.Session() as sess:
    loader.load(sess, tags, saved_model_dir)
    outputs = sess.run('groupwise_dnn_v2/accumulate_scores/truediv:0', feed_dict=inputs_feed_dict)
    
    print(outputs)
```
```
[[0.72908896]
 [0.73187715]
 [0.39144227]]
```

Or in Java:

```
import org.tensorflow.Graph;
import org.tensorflow.Session;
import org.tensorflow.Tensor;
import org.tensorflow.Tensors;
import org.tensorflow.TensorFlow;
import org.tensorflow.SavedModelBundle;
import org.tensorflow.SavedModelBundle.Loader;
import org.tensorflow.framework.SignatureDef;
import org.tensorflow.framework.MetaGraphDef;
import org.tensorflow.framework.TensorInfo;
import org.tensorflow.example.*;
import java.util.Map;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;  

public class HelloTensorFlow {
  
  public static void main(String[] args) throws Exception {
    
      SavedModelBundle bundle = SavedModelBundle.load("., "serve");
      
      Graph graph = bundle.graph();
      
      int cnt=2; //generate cnt fake search query documents to score
      byte[][] examples = new byte[cnt][];
      for(int i=0; i<cnt; i++){      
        Int64List int64List = Int64List.newBuilder().addValue(i).build();
        Feature offset = Feature.newBuilder().setInt64List(int64List).build();
      
        Features features = Features.newBuilder()
                  .putFeature("iid", offset)
                  .build();
        Example example = Example.newBuilder().setFeatures(features).build();
        examples[i]=example.toByteArray();
      }
      
      Tensor<String> inputBatch = Tensors.create(examples);

      final String input_tensor = "input_example_tensor";
      final String output_tensor = "groupwise_dnn_v2/accumulate_scores/truediv";

      Tensor<Float> result = bundle.session().runner()
                      .feed(input_tensor, inputBatch)
                      .fetch(output_tensor)
                      .run().get(0)
                      .expect(Float.class);

      System.out.println(Arrays.deepToString(result.copyTo(new float[cnt][1])));   
  }
}
```
```
STRING tensor with shape [2]
[[0.72908896], [0.73187715]]
```

**Comment by nsalerni** [comment]
Thanks for the quick response! However, even utilizing that examples I still face the following error:

```KeyError: NumericColumn(key='1', shape=(1,), default_value=None, dtype=tf.int64, normalizer_fn=None)```

**Comment by nsalerni** [comment]
The following code snippet solved the issue for me. Adding this a cell in the example notebook produces an exported model in SavedModel format.

```
## Export Model
def example_feature_spec():
    """Returns the example feature columns."""
    feature_names = [
        "%d" % (i + 1) for i in range(0, _NUM_FEATURES)
    ]
    return {
        name: tf.FixedLenFeature([1], tf.float32) for name in feature_names
    }


feature_spec = example_feature_spec()
serving_input_receiver_fn = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
ranker.export_savedmodel('export', serving_input_receiver_fn)
```

**Comment by working-estimate** [comment]
can you share an example of how this works with `build_raw_serving_input_receiver_fn`? The purpose here is to interface with requests that feed features directly instead of through a tf.Example. 

```
def serving_fn():
    features={str(k):tf.placeholder(shape=[None],dtype=tf.float32) for k in range(_NUM_FEATURES+1)}
    return tf.estimator.export.build_raw_serving_input_receiver_fn(features)()
ranker.export_saved_model(export_dir, serving_fn)
``` 
gives the error:
`ValueError: slice index 1 of dimension 0 out of bounds. for 'groupwise_dnn_v2/infer_sizes/strided_slice_1' (op: 'StridedSlice') with input shapes: [1], [1], [1], [1] and with computed input tensors: input[1] = <1>, input[2] = <2>, input[3] = <1>.`

**Comment by mayank953** [comment]
Hi @eggie5,

Thanks for providing with the SavedModel Format, The process works according to the example provided by you. 
I was trying to pass my input in Libsvm format as shown in the example/tf_ranking_libsvm.ipynb with multiple hits of the different query but I was getting an error in doing that.
Do i have to make a new serving function for that?

**Comment by eggie5** [comment]
@mayank953 Since we used `tf.feature_column.make_parse_example_spec` that means the model will expect a `tf.Example` protobuf. 

You can also see this from the output of the saved_model_cli:

```
The given SavedModel SignatureDef contains the following input(s):
  inputs['examples'] tensor_info:
      dtype: DT_STRING
      shape: (-1)
      name: input_example_tensor:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['output'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, -1)
      name: groupwise_dnn_v2/accumulate_scores/truediv:0
Method name is: tensorflow/serving/predict
```

If you want to pass in libsvm at run-time you'll have to define a custom signature receiver function for tensorflow.

**Comment by mayank953** [comment]
@eggie5 Thanks. Actually, I was able to confirm the same. 
Will I have to just change/define my serving input function so as to make it rank my queries? Actually, I was able to predict by the way I worked upon in #69  with @prakhar2811. My input is in that was of format. 

`1 qid:2 2:0.6455186304128903 3:0.5614035087719298 4:74.1518578352181 5:0.8569 6:1440 8:1 10:1 11:1 14:1 15:1
0 qid:2 2:0.634088200238379 3:0.7684210526315789 4:77.24058543002796 5:1.1456 6:1384 8:1 10:1 11:1 14:1 15:1
0 qid:2 2:0.7927631578947368 3:0.5365853658536586 4:51.74825174825175 5:2.0573 6:1380 10:1 11:1 14:1 15:1
0 qid:2 2:0.5254901960784314 3:0.4722222222222222 4:71.91286414024233 5:1.342 6:2179 8:1 11:1 14:1 15:1
0 qid:2 2:0.653179190751445 3:0.8043478260869565 4:65.00159083678015 5:1.3202 6:1100 8:1 11:1 14:1 15:1
0 qid:2 2:0.6559714795008913 3:0.8224299065420561 4:65.80720773513039 5:2.5779 6:2334 8:1 10:1 11:1 14:1 15:1
0 qid:2 2:0.6503496503496503 4:50 5:3.4414 6:639 11:1 14:1 15:1
0 qid:2 2:0.7243150684931506 3:0.5819672131147541 4:76.81513117754729 5:3.1104 6:1140 8:1 10:1 11:1 14:1 15:1
0 qid:2 2:0.7222222222222222 3:0.3333333333333333 4:62.51140164183642 5:0.4645 6:1233 8:1 11:1 14:1 15:1
0 qid:2 2:0.5888888888888889 4:53.02663438256658 5:2.4784 6:776 11:1 14:1 15:1
1 qid:3 4:25.00846023688663 5:1.2924 6:6648 10:1 14:1 15:1
1 qid:3 2:0.5155172413793103 3:0.4923076923076923 4:25.00593119810202 5:2.1348 6:3161 8:1 10:1 14:1 15:1
0 qid:3 2:0.7123287671232876 3:0.3771929824561404 4:25.01059771089445 5:0.3972 6:3538 14:1 15:1
0 qid:3 2:0.6503496503496503 4:25 5:1.9369 6:6750 14:1 15:1
0 qid:3 2:0.9 4:25.01652346331791 5:1.1211 6:2269 8:1 14:1 15:1
1 qid:4 2:0.6455186304128903 3:0.5614035087719298 4:70.69440195778526 5:0.8569 6:1916 8:1 11:1 14:1 15:1`

Basically, it's a LibSvm format with multiple entries for qid 3,4&5 for which I was able to predict the new rank & improved NDCG by using normal ranker. predict but was not I want to predict it using saved_model_cli run.
Thanks

**Comment by hodeld** [comment]
Can you share an example how to save and use the trained model used in the tutorial [handling_sparse_features](https://colab.research.google.com/github/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb#scrollTo=cROuS7-8Dd-r)?
I would like to make predictions as in the example but from a saved model. I don't see how to export with `ranker.export_savedmodel()` as in the tutorial there are 2 kind of feature_spec (examples and context). So I would like to define a `build_parsing_serving_input_receiver_fn` similar to the `predict_input_fn(path)` in the tutorial.

```
def predict_input_fn(path):
  context_feature_spec = tf.feature_column.make_parse_example_spec(
        context_feature_columns().values())
  example_feature_spec = tf.feature_column.make_parse_example_spec(
        list(example_feature_columns().values()))
  dataset = tfr.data.build_ranking_dataset(
        file_pattern=path,
        data_format=tfr.data.ELWC,
        batch_size=_BATCH_SIZE,
        list_size=_LIST_SIZE,
        context_feature_spec=context_feature_spec,
        example_feature_spec=example_feature_spec,
        reader=tf.data.TFRecordDataset,
        shuffle=False,
        num_epochs=1)
  features = tf.compat.v1.data.make_one_shot_iterator(dataset).get_next()
  return features
```
Thanks!

**Comment by kunalkishore** [comment]
@hodeld This might help you:

```
def make_serving_input_fn():
    """Returns serving input fn."""
    context_feature_spec = tf.feature_column.make_parse_example_spec(
      context_feature_columns().values())
    example_feature_spec = tf.feature_column.make_parse_example_spec(
      example_feature_columns().values())
    return tfr.data.build_ranking_serving_input_receiver_fn(
        data_format="example_list_with_context",
        context_feature_spec=context_feature_spec,
        example_feature_spec=example_feature_spec)

estimator_path = ranker.export_saved_model("path/to/export", make_serving_input_fn())
```
